File: backend/app/domain/auth/password_policy.py
"""Política de contraseñas del sistema (RN-AU).

Nivel **Media** (decisión del dueño): mínimo 8 caracteres, con al menos una
mayúscula, una minúscula y un dígito. Aplica a:
  - auto-registro de estudiantes (`POST /auth/register`),
  - cambio de contraseña propio (`PUT /auth/change-password`),
  - definición de la contraseña definitiva tras el primer login (clave temporal).

Función pura, sin dependencias de DB ni framework: fácil de testear y reutilizar.
"""

from __future__ import annotations

LONGITUD_MINIMA = 8


class PasswordDebilError(ValueError):
    """La contraseña no cumple la política. El mensaje lista los requisitos faltantes."""


def _requisitos_incumplidos(password: str) -> list[str]:
    faltantes: list[str] = []
    if len(password) < LONGITUD_MINIMA:
        faltantes.append(f"al menos {LONGITUD_MINIMA} caracteres")
    if not any(c.isupper() for c in password):
        faltantes.append("una letra mayúscula")
    if not any(c.islower() for c in password):
        faltantes.append("una letra minúscula")
    if not any(c.isdigit() for c in password):
        faltantes.append("un número")
    return faltantes


def validar_password_fuerte(password: str) -> None:
    """Valida la contraseña contra la política Media. Lanza ``PasswordDebilError`` si no cumple.

    El mensaje enumera TODOS los requisitos faltantes para que el usuario los
    corrija de una sola vez.
    """
    faltantes = _requisitos_incumplidos(password)
    if faltantes:
        raise PasswordDebilError(
            "La contraseña debe tener " + ", ".join(faltantes) + "."
        )

File: frontend/src/lib/passwordPolicy.ts
/**
 * Política de contraseñas del frontend — espeja `backend/app/domain/auth/password_policy.py`.
 *
 * Nivel Media: mínimo 8 caracteres, con al menos una mayúscula, una minúscula y
 * un dígito. Valida en cliente para dar feedback inmediato; el backend vuelve a
 * validar (nunca se confía en el cliente).
 */

export const LONGITUD_MINIMA_PASSWORD = 8;

export const REQUISITOS_PASSWORD = 'Mínimo 8 caracteres, con una mayúscula, una minúscula y un número.';

/**
 * Devuelve un mensaje con los requisitos faltantes, o `null` si la contraseña
 * cumple la política.
 */
export function validarPasswordFuerte(password: string): string | null {
  const faltantes: string[] = [];
  if (password.length < LONGITUD_MINIMA_PASSWORD) faltantes.push(`al menos ${LONGITUD_MINIMA_PASSWORD} caracteres`);
  if (!/[A-Z]/.test(password)) faltantes.push('una letra mayúscula');
  if (!/[a-z]/.test(password)) faltantes.push('una letra minúscula');
  if (!/[0-9]/.test(password)) faltantes.push('un número');
  if (faltantes.length === 0) return null;
  return 'La contraseña debe tener ' + faltantes.join(', ') + '.';
}
